### PR TITLE
Ender 6179  - extend ezdb with foreign key functionality

### PR DESCRIPTION
**Pull request by amogiska**

### CODE REVIEW

File: src/ez/helper/ForeignKeyConstraint.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import ox.Log;
+
+public class ForeignKeyConstraint {
+
+  protected String foreigKeyName, sourceTable, sourceColumnName, foreignTable, foreignColumnName;
+
+  protected ForeignKeyConstraint(String sourceTable, String sourceColumnName, String foreignTable,
+      String foreignColumnName) {
+    this.sourceTable = sourceTable;
```
Comment (mirraj2) [inline src/ez/helper/ForeignKeyConstraint.java:16]: this should call the other constructor

File: src/ez/helper/ForeignKeyConstraint.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    this.sourceColumnName = sourceColumnName;
+    this.foreignTable = foreignTable;
+    this.foreignColumnName = foreignColumnName;
+  }
+
+  protected ForeignKeyConstraint(String sourceTable, String sourceColumnName, String foreignTable,
+      String foreignColumnName, String foreignKeyName) {
+    this.sourceTable = sourceTable;
```
Comment (mirraj2) [inline src/ez/helper/ForeignKeyConstraint.java:24]: this should call normalize() on all arguments.  It should also checkNotEmpty() on once that it expects to have values

File: src/ez/Table.java [hunk 3]
```diff
@@ -98,6 +102,29 @@ public Table varchar(String name, int n) {
     return column(name, "VARCHAR(" + n + ")");
   }
 
+  public Table linkColumn(String name, Class<?> type, String foreignTable, String foreignColumnName,
+      String foreignKeyName) {
```
Comment (mirraj2) [inline src/ez/Table.java:106]: lets just not let people specify foreignKeyName

File: src/ez/impl/MySQLDB.java [hunk 4]
```diff
@@ -89,6 +89,20 @@ public boolean hasColumn(String table, String column) {
         schema, table, column);
   }
 
+  @Override
+  public boolean hasForeignKey(String sourceTable, String sourceColumn, String foreignTable, String foreignColumn) {
+    return null != selectSingleRow("SELECT `COLUMN_NAME`"
+        + " FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ? AND COLUMN_NAME = ? AND REFERENCED_TABLE_NAME = ? AND REFERENCED_COLUMN_NAME = ? LIMIT 1",
```
Comment (mirraj2) [inline src/ez/impl/MySQLDB.java:95]: formatting

**Review by mirraj2 — CHANGES_REQUESTED** [review]

File: src/ez/Table.java (hunk 3 above)
Comment (amogiska) [reply src/ez/Table.java:106]: removed the link column with foreignKeyName

File: src/ez/helper/ForeignKeyConstraint.java (hunk 2 above)
Comment (amogiska) [reply src/ez/helper/ForeignKeyConstraint.java:24]: done

File: src/ez/impl/MySQLDB.java (hunk 4 above)
Comment (amogiska) [reply src/ez/impl/MySQLDB.java:95]: not sure what is wrong with the formatting looks similar to hasColumn and I have the correct formatter on ecllipse for Ender. Is there something separate I have to go for EZDB?

File: src/ez/helper/ForeignKeyBuilder.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  public ForeignKeyBuilder foreignColumn(String foreignColumn) {
+    this.foreignColumnName = foreignColumn;
+    return this;
+  }
+
+  public ForeignKeyConstraint build() {
+    if (Strings.isNullOrEmpty(constraintName)) {
+      return new ForeignKeyConstraint(sourceTable, sourceColumnName, foreignTable, foreignColumnName);
+    } else {
+      checkState(constraintName.length() < 64, "The constraint name is too longer than 64 characters");
```
Comment (mirraj2) [inline src/ez/helper/ForeignKeyBuilder.java:39]: wrong grammar

File: src/ez/impl/MySQLDB.java (hunk 4 above)
Comment (mirraj2) [reply src/ez/impl/MySQLDB.java:95]: lines should not be longer than 120 characters

File: src/ez/impl/MySQLDB.java [hunk 6]
```diff
@@ -89,6 +89,20 @@ public boolean hasColumn(String table, String column) {
...
+        schema, sourceTable, sourceColumn, foreignTable, foreignColumn);
+  }
+
+  @Override
+  public String getForeignKeyName(String sourceTable, String sourceColumn, String foreignTable, String foreignColumn) {
+    return selectSingleRow("SELECT `CONSTRAINT_NAME`"
+        + " FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ? AND COLUMN_NAME = ? AND REFERENCED_TABLE_NAME = ? AND REFERENCED_COLUMN_NAME = ? LIMIT 1",
```
Comment (mirraj2) [inline src/ez/impl/MySQLDB.java:102]: lines should not be longer than 120 characters

**Review by mirraj2 — CHANGES_REQUESTED** [review]

File: src/ez/impl/MySQLDB.java (hunk 4 above)
Comment (amogiska) [reply src/ez/impl/MySQLDB.java]: fixed

**Review by mirraj2 — APPROVED** [review]

**Pull request merged**